frontend/src/breeding_form.tsx

```diff
@@ -0,0 +1,199 @@
+/**
+ * @file This file contains the BreedingForm function. This function allows
+ *       users to create and update breeding events in the database.
+ */
+import React from 'react'
+import { makeStyles } from '@material-ui/core/styles';
+import { MuiPickersUtilsProvider, KeyboardDatePicker
+        } from '@material-ui/pickers'
+import DateFnsUtils from '@date-io/date-fns'
+import { dateFormat, Genebank, individualLabel, inputVariant, LimitedHerd,
+        OptionType } from '@app/data_context_global';
+
+import { Breeding } from './breeding_list';
+import { TextField, Typography } from '@material-ui/core';
+import { useDataContext } from './data_context';
+import { Autocomplete } from '@material-ui/lab';
+
+const useStyles = makeStyles({
+  form: {
+    width: '100%',
+    height: '100%',
+    overflow: 'hidden',
+    padding: '20px',
+  },
+  formBox: {
+    border: '1px solid lightgrey',
+    borderRadius: '8px',
+    padding: '10px',
+  },
+  wideControl: {
+    width: '100%',
+  },
+});
+
+const emptyBreeding: Breeding = {
+  breed_date: '',
+  breed_notes: '',
+  father: '',
+  mother: '',
+  birth_date: '',
+  birth_notes: '',
+  litter_size: 0,
+}
+
+/**
+ * The BreedingForm function. This function allows users to create and update
+ * breeding events in the database.
+ */
+export function BreedingForm({data, herdId}: {data: Breeding | 'new', herdId: string | undefined}) {
+  const style = useStyles();
+  const { genebanks } = useDataContext()
+  const [form, setForm] = React.useState({...emptyBreeding} as Breeding)
```

Comment:
There is no point in copying the `emptyBreeding` object to the state. `emptyBreeding` should stay constant, it's a bug if this object is ever modified. (If you expect this happens you can debug it by using Object.freeze and Object.seal on emptyBreeding to make an error be thrown in this happens)

On the other hand, values passed to `setForm` shoud be copies, not mutated objects referenced elsewhere. Because if this would ever happen it could mean that you could make emptyBreeding get mutated.

Futher, the variable name `form` is a bit unfortunate because it's not the actual form, it's the state in the from. Up to you of course but I would go for `form_state` or simply `state` (because it's the main state of this component).